File: scripts/03_inferential.R
# ============================================================
# IT3011 Group Assignment - Script 03: Inferential Statistics
# Learning Outcomes: LO2 - Hypothesis Testing
# ============================================================

library(tidyverse)

# ---- 1. Load Cleaned Data ----
cat("Loading cleaned dataset...\n")
data <- read.csv("data/processed/cleaned_data.csv")
data$SelfAssessment_Level <- factor(data$SelfAssessment_Level, levels = c("Low", "Moderate", "High"))

# ---- 2. Define Hypotheses ----
# H0: There is no significant correlation between Assignment Completion (Self-assessment proxy) and Exam Score.
# H1: There is a significant positive correlation between Assignment Completion and Exam Score.

# H0_2: There is no significant difference in Exam Scores across different Self-Assessment Levels.
# H1_2: There is a significant difference in Exam Scores across different Self-Assessment Levels.

# ---- 3. Check Assumptions ----
cat("\n--- Checking Assumptions ---\n")
# Normality test (using a sample if n > 5000, shapiro.test requires n >= 3 && n <= 5000)
# We will use a random sample of 4000 to test normality if dataset is large
sample_data <- data
if(nrow(data) > 5000) { sample_data <- data %>% sample_n(4000) }

cat("Shapiro-Wilk Test for Normality (Exam Score):\n")
shapiro_res <- shapiro.test(sample_data$ExamScore)
print(shapiro_res)

cat("\nBartlett's Test for Homogeneity of Variance:\n")
bartlett_res <- bartlett.test(ExamScore ~ SelfAssessment_Level, data = data)
print(bartlett_res)

# ---- 4. Run Tests ----
cat("\n--- Running Inferential Tests ---\n")

# Correlation Test
cat("Pearson Correlation Test:\n")
cor_res <- cor.test(data$AssignmentCompletion, data$ExamScore, method = "pearson")
print(cor_res)

# ANOVA Test
cat("\nOne-Way ANOVA (Exam Score by Self-Assessment Level):\n")
anova_res <- aov(ExamScore ~ SelfAssessment_Level, data = data)
print(summary(anova_res))

# Post-hoc test if ANOVA is significant
if (summary(anova_res)[[1]][["Pr(>F)"]][1] < 0.05) {
  cat("\nANOVA is significant. Running Tukey HSD post-hoc test:\n")
  tukey_res <- TukeyHSD(anova_res)
  print(tukey_res)
}

# ---- 5. Save Output ----
dir.create("output/tables", recursive = TRUE, showWarnings = FALSE)
# Using sink to save printed output to a text file for documentation
sink("output/tables/inferential_results.txt")
cat("Correlation Test:\n")
print(cor_res)
cat("\nANOVA Summary:\n")
print(summary(anova_res))
if (exists("tukey_res")) {
  cat("\nTukey HSD:\n")
  print(tukey_res)
}
sink()

cat("============================================================\n")
cat("Inferential analysis complete. Results saved in output/tables/\n")
cat("Proceed to 04_predictive.R\n")
cat("============================================================\n")

File: scripts/03_inferential.py
import pandas as pd
import scipy.stats as stats
import os

try:
    import statsmodels.api as sm
    from statsmodels.stats.multicomp import pairwise_tukeyhsd
    HAS_STATSMODELS = True
except ImportError:
    HAS_STATSMODELS = False
    print("Warning: statsmodels not installed. Tukey HSD will be skipped.")

print("============================================================")
print("IT3011 Group Assignment - Script 03: Inferential stats (Python)")
print("Learning Outcomes: LO2")
print("============================================================\n")

# ---- 1. Load Cleaned Data ----
print("Loading cleaned dataset...")
try:
    data = pd.read_csv("data/processed/cleaned_data.csv")
    data['SelfAssessment_Level'] = pd.Categorical(data['SelfAssessment_Level'], categories=["Low", "Moderate", "High"], ordered=True)
except FileNotFoundError:
    print("Cleaned data not found. Run 01_data_cleaning.py first.")
    exit(1)

# ---- 3. Check Assumptions ----
print("\n--- Checking Assumptions ---")
# Normality test
sample_data = data.sample(4000) if len(data) > 5000 else data
shapiro_stat, shapiro_p = stats.shapiro(sample_data['ExamScore'])
print(f"Shapiro-Wilk Test (Exam Score): W={shapiro_stat:.4f}, p-value={shapiro_p:.4e}")

# Homogeneity of variance (Levene or Bartlett)
groups = [data[data['SelfAssessment_Level'] == level]['ExamScore'] for level in data['SelfAssessment_Level'].cat.categories]
bartlett_stat, bartlett_p = stats.bartlett(*groups)
print(f"Bartlett's Test for Homogeneity of Variance: stat={bartlett_stat:.4f}, p-value={bartlett_p:.4e}")

# ---- 4. Run Tests ----
print("\n--- Running Inferential Tests ---")

# Correlation Test
pearson_corr, pearson_p = stats.pearsonr(data['AssignmentCompletion'], data['ExamScore'])
print(f"Pearson Correlation Test: r={pearson_corr:.4f}, p-value={pearson_p:.4e}")

# ANOVA Test
f_stat, anova_p = stats.f_oneway(*groups)
print(f"\nOne-Way ANOVA (Exam Score by Self-Assessment Level): F={f_stat:.4f}, p={anova_p:.4e}")

# Tukey HSD
tukey_res_str = ""
if anova_p < 0.05 and HAS_STATSMODELS:
    print("\nANOVA is significant. Running Tukey HSD post-hoc test:")
    tukey_res = pairwise_tukeyhsd(endog=data['ExamScore'], groups=data['SelfAssessment_Level'], alpha=0.05)
    print(tukey_res)
    tukey_res_str = str(tukey_res)

# ---- 5. Save Output ----
os.makedirs("output/tables", exist_ok=True)
with open("output/tables/inferential_results_py.txt", "w") as f:
    f.write(f"Pearson Correlation Test: r={pearson_corr:.4f}, p={pearson_p:.4e}\n")
    f.write(f"ANOVA Test: F={f_stat:.4f}, p={anova_p:.4e}\n")
    if tukey_res_str:
        f.write("\nTukey HSD:\n")
        f.write(tukey_res_str + "\n")

print("\nInferential analysis complete. Results saved in output/tables/")
